src/websocket_tracker_connection.cpp

```diff
@@ -0,0 +1,490 @@
+    payload["action"] = "announce";
+    payload["info_hash"] = from_latin1(ans.info_hash);
+    payload["offer_id"] = from_latin1(ans.answer.offer_id);
+    payload["to_peer_id"] = from_latin1(ans.answer.pid);
+    payload["peer_id"] =  from_latin1(ans.pid);
+	json::object &obj = payload["answer"].emplace_object();
+    obj["type"] = "answer";
+    obj["sdp"] = ans.answer.sdp;
+
+	json::string const data = json::to_string(payload);
+
+#ifndef TORRENT_DISABLE_LOGGING
+	if(auto cb = requester())
+		cb->debug_log("*** WEBSOCKET_TRACKER_WRITE [ size: %d, data: %s ]", int(data.size()), data.c_str());
+#endif
+
+	ADD_OUTSTANDING_ASYNC("websocket_tracker_connection::on_write");
+	m_websocket->async_write(boost::asio::const_buffer(data.data(), data.size())
+			, std::bind(&websocket_tracker_connection::on_write, this, weak_from_this(), _1, _2));
+}
+
+void websocket_tracker_connection::do_read()
+{
+	if(!is_open()) return;
+
+	// Can be replaced by m_read_buffer.clear() with boost 1.70+
+	if(m_read_buffer.size() > 0) m_read_buffer.consume(m_read_buffer.size());
+
+	ADD_OUTSTANDING_ASYNC("websocket_tracker_connection::on_read");
+	m_websocket->async_read(m_read_buffer
+            , std::bind(&websocket_tracker_connection::on_read, this, weak_from_this(), _1, _2));
+}
+
+void websocket_tracker_connection::on_timeout(error_code const& ec)
+{
+	// No async
+	if(ec)
+    {
+		fail(operation_t::sock_read, ec);
+		return;
+	}
+
+#ifndef TORRENT_DISABLE_LOGGING
+	if(auto cb = requester())
+		cb->debug_log("*** WEBSOCKET_TRACKER_TIMEOUT [ url: %s ]", tracker_req().url.c_str());
+#endif
+
+	fail(operation_t::sock_read, error_code(error::timed_out));
+	close();
+}
+
+void websocket_tracker_connection::on_connect(std::weak_ptr<websocket_tracker_connection> weak_this
+		, error_code const& ec)
+{
+	COMPLETE_ASYNC("websocket_tracker_connection::on_connect");
+	auto locked = weak_this.lock();
+	if(!locked) return;
+
+	if(ec)
+	{
+		fail(operation_t::connect, ec);
+		close();
+		return;
+	}
+
+	send_pending();
+	do_read();
+}
+
+void websocket_tracker_connection::on_read(std::weak_ptr<websocket_tracker_connection> weak_this
+		, error_code const& ec, std::size_t /* bytes_read */)
+{
+	COMPLETE_ASYNC("websocket_tracker_connection::on_read");
+	auto locked = weak_this.lock();
+	if(!locked) return;
+
+	if(ec)
+	{
+		fail(operation_t::sock_read, ec);
+		close();
+		return;
+	}
+
+	std::shared_ptr<request_callback> cb = requester();
+    try {
+    	auto const& buf = m_read_buffer.data();
+		auto const data = static_cast<char const*>(buf.data());
+    	auto const size = buf.size();
+
+#ifndef TORRENT_DISABLE_LOGGING
+		std::string str(data, size);
+		if(cb) cb->debug_log("*** WEBSOCKET_TRACKER_READ [ size: %d, data: %s ]", int(str.size()), str.c_str());
+#endif
+
+		json::object payload = json::parse({data, size}).as_object();
+
+		auto it = payload.find("info_hash");
+		if(it == payload.end())
+			throw std::invalid_argument("no info hash in message");
```

Comment:
I think the tracker infrastructure expects a call to `fail()` rather than an exception